File: JavaPractice/src/LeetCode79WordSearchDFS.java
/**
 * @author divyesh.shah
 * Given an m x n grid of characters board and a string word, return true if word exists in the grid.
 *
 * The word can be constructed from letters of sequentially adjacent cells, where adjacent cells are horizontally or vertically neighboring. The same letter cell may not be used more than once.
 *
 * e.g1 - Input: board = [["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]], word = "ABCCED"
 * Output: true
 *
 * e.g2 - Input: board = [["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]], word = "SEE"
 * Output: true
 *
 * e.g3 - Input: board = [["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]], word = "ABCB"
 * Output: false
 *
 * Constraints:
 *
 * m == board.length
 * n = board[i].length
 * 1 <= m, n <= 6
 * 1 <= word.length <= 15
 * board and word consists of only lowercase and uppercase English letters.
 *
 *
 * Follow up: Could you use search pruning to make your solution faster with a larger board?
 */
public class LeetCode79WordSearchDFS {

    public static void main(String[] args) {
        char [][] c = new char[][]{{'A','B','C','E'},{'S','F','C','S'},{'A','D','E','E'}};
        System.out.println(exist(c,"ABCCED"));
        System.out.println(exist(c,"SEE"));
        System.out.println(exist(c,"ABCB"));
    }

    public static boolean exist(char[][] board, String word) {
        if (board == null || board.length == 0 || board[0].length == 0 || word == null || word.length() == 0) {
            return false;
        }
        int m = board.length;
        int n = board[0].length;
        boolean[][] visited = new boolean[m][n];
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (dfs(board, visited, i, j, word, 0)) {
                    return true;
                }
            }
        }
        return false;
    }

    private static boolean dfs(char[][] board, boolean[][] visited, int i, int j, String word, int k) {
        if (k == word.length()) {
            return true;
        }
        if (i < 0 || i >= board.length || j < 0 || j >= board[0].length || visited[i][j] || board[i][j] != word.charAt(k)) {
            return false;
        }
        visited[i][j] = true;
        boolean res = dfs(board, visited, i + 1, j, word, k + 1) ||
                dfs(board, visited, i - 1, j, word, k + 1) ||
                dfs(board, visited, i, j + 1, word, k + 1) ||
                dfs(board, visited, i, j - 1, word, k + 1);
        visited[i][j] = false;
        return res;
    }
}

File: PythonPractice/src/LeetCode79WordSearchDFS.py
from typing import List, Optional

def exist(board: Optional[List[List[str]]], word: Optional[str]) -> bool:
    if board is None or len(board) == 0 or len(board[0]) == 0 or word is None or len(word) == 0:
        return False

    def dfs(i: int, j: int, k: int) -> bool:
        if k == len(word):
            return True
        if (i < 0 or i >= len(board) or
                j < 0 or j >= len(board[0]) or
                visited[i][j] or board[i][j] != word[k]):
            return False

        visited[i][j] = True
        res = (dfs(i + 1, j, k + 1) or
               dfs(i - 1, j, k + 1) or
               dfs(i, j + 1, k + 1) or
               dfs(i, j - 1, k + 1))
        visited[i][j] = False
        return res

    m = len(board)
    n = len(board[0])
    visited = [[False] * n for _ in range(m)]

    for i in range(m):
        for j in range(n):
            if dfs(i, j, 0):
                return True

    return False

# Example usage
if __name__ == "__main__":
    board = [['A', 'B', 'C', 'E'],
             ['S', 'F', 'C', 'S'],
             ['A', 'D', 'E', 'E']]

    print(exist(board, "ABCCED"))  # Output: True
    print(exist(board, "SEE"))     # Output: True
    print(exist(board, "ABCB"))    # Output: False
